src/libraries/Hooks.sol

```diff
@@ -259,27 +258,30 @@ library Hooks {
         IPoolManager.SwapParams memory params,
         BalanceDelta swapDelta,
         bytes calldata hookData,
-        int128 hookDeltaSpecified
-    ) internal returns (BalanceDelta swapperDelta, BalanceDelta hookDelta) {
+        HookReturnDelta beforeSwapHookReturn
+    ) internal returns (BalanceDelta, BalanceDelta) {
```

Comment:
Can you elaborate why `beforeSwap` can modify both `specified` and `unspecified` balances, while `afterSwap` can only modify the `unspecified` balance?

What are some of the possible scenarios where `afterSwap` only modify `unspecified` while `beforeSwap` modified both? Will it be better off if both before/after can modify both?

Reply:
In addition, is there a way to know the unspecified delta amount in the `beforeSwap` hook? If i need to take a portion (say 0.1%) of the unspecified, is it possible to do it in the beforeSwap hook?

Reply:
hi! i think theres a few key points to explain this:
- the `specified` amount is the amount that defines the exact amount that a user wants to swap - whether input or output. No matter what the hook returns in the `beforeSwap` hook to alter the amount actually swapped on the pool, the user will see a difference of `amountSpecified` in their wallet to maintain expected behaviour.
- After the swap has happened, its therefore impossible to alter this `specified` amount because its already been swapped on the pool and that cannot be reversed. That is why there is no `deltaSpecified` in the `afterSwap` hook. `deltaSpecified` is to alter how much is to be swapped on the pool - so in `afterSwap` it is too late
- For `deltaUnspecified` there are 2 main scenarios:
1. Hooks wanting to take a % (or some formula) of the unspecified currency of a trade to take a fee. The unspecified amount of a swap on a pool is not known until the actual swap has taken place (unless you simulate the full swap and liquidity amounts to calculate it elsewhere). Hooks like this will want to therefore apply this delta in the `afterSwap` hook as they cannot know how much to take until after the swap has taken place
2. Hooks that implement a "custom curve". This means that the hooks actually do not allow swaps to happen in the V4 Concentrated Liquidity pool and instead implement custom swap logic internal to the hook. These hooks will return `-amountSpecified` as their `specified` delta, so that the V4 pool will trade `amountSpecified-amountSpecified = 0` on the V4 pool, aka no swap occurs. Instead they then calculate their own custom unspecified currency amount for the trade. These hooks _do not_ need to wait for `afterSwap` hook to know their `deltaSpecified` because their formula is independent of the swap logic in the pool manager. Before this pull request such hooks had to return this deltaUnspecified in the `afterSwap` hook which is unnecessarily inefficient. This PR lets those hooks return both values in `beforeSwap`.

I do not envision there being a good reason for a hook to need to return a `deltaUnspecified` in _both_ the before and after swap hooks. Most should just choose 1 - depending if their deltaUnspecified is independent of the unspecified of the swap on V4, or not. However it is technically possible to return both because the code does not stop it.

Reply:
Does that make sense? 😅 I get this question a lot so definitely want to find an explanation that makes sense for us to put in the docs haha

Reply:
Thanks for the detailed explanation! This makes lots of sense!

What I am trying to do is to define a `preferred currency` to take `hookFee` on. So this `preferred currency` could either be a `specified currency` or an `unspecified currency`. I think in that case I will have to have both before&after hooks, and then depends on the direction and the token specified to decide where to take the hook fee.

Reply:
If your fee is a % of the input (or output) then yes I think youll need to implement both hooks. However you could save double-calculating things by saving intermediary results in beforeSwap in transient storage to then reference in the afterSwap hook :)